### ISSUE TITLE
Mouse Event

### ISSUE THREAD

**Issue by noisewuwei** [ocornut/imgui#5203]
OS: Mac OS 
Backend:Metal
I created two NSWindow with MTKView. When my imgui window name is the same, two NSWindows can respond to mouse events on any one window at the same time. When imgui window names are different, two NSWindows cannot respond to mouse events. What caused it?

**Comment by ocornut** [comment]
Hello,
It is ambiguous whether you created multiple imgui contexts or not, or what you do you mean by "my imgui window", etc. I think you will need to provide more details in order to receive help. Please post more details and I will reopen this issue.

**Comment by ocornut** [comment]
If you create multiple contexts you need to switch context with `ImGui::SetCurrentContext()` before accessing the IO structure.
It is unclear which platform backend you are using (you didn't fill the requested template).

**Comment by noisewuwei** [comment]
> If you create multiple contexts you need to switch context with `ImGui::SetCurrentContext()` before accessing the IO structure.
> It is unclear which platform backend you are using (you didn't fill the requested template).

MAC OS  Metal.

**Comment by noisewuwei** [comment]
@interface IMGUIMetalView ()<MTKViewDelegate>{
    ImGuiContext *ctx;
}

- (void)drawRect:(NSRect)dirtyRect {
    [super drawRect:dirtyRect];
    // Drawing code here.
     ctx = ImGui::CreateContext(); 
     ImGuiIO& io = ImGui::GetIO(); (void)io;
}

-(void)drawInMTKView:(MTKView*)view
{    
    ImGui::SetCurrentContext(ctx);
    
    ImGuiIO& io = ImGui::GetIO();
    io.DisplaySize.x = view.bounds.size.width;
    io.DisplaySize.y = view.bounds.size.height;
    
#if TARGET_OS_OSX
    CGFloat framebufferScale = view.window.screen.backingScaleFactor ?: NSScreen.mainScreen.backingScaleFactor;
#else
    CGFloat framebufferScale = view.window.screen.scale ?: UIScreen.mainScreen.scale;
#endif
    io.DisplayFramebufferScale = ImVec2(framebufferScale, framebufferScale);

    io.DeltaTime = 1 / float(view.preferredFramesPerSecond ?: 60);
}

Is that right？

**Comment by noisewuwei** [comment]
> If you create multiple contexts you need to switch context with `ImGui::SetCurrentContext()` before accessing the IO structure.
> It is unclear which platform backend you are using (you didn't fill the requested template).

[https://github.com/ocornut/imgui/commit/4cec3a0467af9cd068d85363d1a7ceaddc3b68e0](url)

I've seen this link.
Metal cannot support multi-context ？

**Comment by ocornut** [comment]
Right, neither Metal nor OSX backend have been refactored to support multiple context.

**Comment by noisewuwei** [comment]
> Right, neither Metal nor OSX backend have been refactored to support multiple context.

Metal cannot support multiple-viewprots neither.Is there any way to create multiple windows when using Metal on Macos.

**Comment by ocornut** [comment]
Well we need pull requests for both.
Multi-contexts should be quite easy to support. Multi-viewport more involved.

**Comment by noisewuwei** [comment]
> Well we need pull requests for both.
> Multi-contexts should be quite easy to support. Multi-viewport more involved.

Yep.Multi-viewports is like magic.

**Comment by ocornut** [comment]
I don't work for Mac so it depends on other users.

For multi-viewport see #2778 and #4821

For multi-contexts, we need both OSX and Metal backends to the same as e.g #4141 with 70c603854e8817770c1a4eeb047929eb1b7313a3 and 4cec3a0467af9cd068d85363d1a7ceaddc3b68e0
This probably needs to be done first before the multi-viewports stuff gets merged (then #2778 and #4821 rebased over that)

**Comment by noisewuwei** [comment]
> I don't work for Mac so it depends on other users.
> 
> For multi-viewport see #2778 and #4821
> 
> For multi-contexts, we need both OSX and Metal backends to the same as e.g #4141 with [70c6038](https://github.com/ocornut/imgui/commit/70c603854e8817770c1a4eeb047929eb1b7313a3) and [4cec3a0](https://github.com/ocornut/imgui/commit/4cec3a0467af9cd068d85363d1a7ceaddc3b68e0) This probably needs to be done first before the multi-viewports stuff gets merged (then #2778 and #4821 rebased over that)

OK.I will have a try.

**Comment by noisewuwei** [comment]
> For multi-contexts, we need both OSX and Metal backends to the same
[muti-contexts.zip](https://github.com/ocornut/imgui/files/8519482/muti-contexts.zip)

Hello.I've refactored the code in imgui_impl_metal.mm and imgui_impl_osx.mm.I have upload two source code files.But the program crashed when I create the second context.
error is : (bd->sharedMetalContext != nullptr && "No Metal context. Did you call ImGui_ImplMetal_Init() ?"), function ImGui_ImplMetal_NewFrame, file imgui_impl_metal.mm, line 164.

What should I do?

**Comment by noisewuwei** [comment]
> I don't work for Mac so it depends on other users.
> 
> For multi-viewport see #2778 and #4821
> 
> For multi-contexts, we need both OSX and Metal backends to the same as e.g #4141 with [70c6038](https://github.com/ocornut/imgui/commit/70c603854e8817770c1a4eeb047929eb1b7313a3) and [4cec3a0](https://github.com/ocornut/imgui/commit/4cec3a0467af9cd068d85363d1a7ceaddc3b68e0) This probably needs to be done first before the multi-viewports stuff gets merged (then #2778 and #4821 rebased over that)

Hello.I've refactored the code in imgui_impl_metal.mm and imgui_impl_osx.mm.I have upload two source code files.But the program crashed when I create the second context.
error is : (bd->sharedMetalContext != nullptr && "No Metal context. Did you call ImGui_ImplMetal_Init() ?"), function ImGui_ImplMetal_NewFrame, file imgui_impl_metal.mm, line 164.

What should I do?
[muti-contexts.zip](https://github.com/ocornut/imgui/files/8519546/muti-contexts.zip)

**Comment by noisewuwei** [comment]
> I don't work for Mac so it depends on other users.
> 
> For multi-viewport see #2778 and #4821
> 
> For multi-contexts, we need both OSX and Metal backends to the same as e.g #4141 with [70c6038](https://github.com/ocornut/imgui/commit/70c603854e8817770c1a4eeb047929eb1b7313a3) and [4cec3a0](https://github.com/ocornut/imgui/commit/4cec3a0467af9cd068d85363d1a7ceaddc3b68e0) This probably needs to be done first before the multi-viewports stuff gets merged (then #2778 and #4821 rebased over that)

[https://github.com/ocornut/imgui/pull/5221](url)

Hello.Take a look.

**Comment by noisewuwei** [comment]
> I don't work for Mac so it depends on other users.
> 
> For multi-viewport see #2778 and #4821
> 
> For multi-contexts, we need both OSX and Metal backends to the same as e.g #4141 with [70c6038](https://github.com/ocornut/imgui/commit/70c603854e8817770c1a4eeb047929eb1b7313a3) and [4cec3a0](https://github.com/ocornut/imgui/commit/4cec3a0467af9cd068d85363d1a7ceaddc3b68e0) This probably needs to be done first before the multi-viewports stuff gets merged (then #2778 and #4821 rebased over that)

Hello.
#5221 pull request.
Take a look.

**Comment by ocornut** [comment]
Thank you! We have now merged e54db4e which is your PR + various fixes/cleanup done by @rokups. 
(Your PR accidentally reverted the two latest commit on those files)
